File: src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"
using namespace std::placeholders;
using namespace std::chrono_literals;

class NumberCounterNode : public rclcpp::Node 
{
public:
    NumberCounterNode() : Node("number_counter"), counter_(0.0)
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10 , std::bind(&NumberCounterNode::callbackNumberCounter, this, _1)
        );
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count",10);
        timer_ = this->create_wall_timer(0.5s,std::bind(&NumberCounterNode::publishCounter,this));


        server_ = this->create_service<example_interfaces::srv::SetBool>("reset_counter", std::bind(&NumberCounterNode::callBackResetCounter,this, _1,_2));

        RCLCPP_INFO(this->get_logger(), "Number counter has been started.");
    }
 
private:
    void callbackNumberCounter(const example_interfaces::msg::Int64::SharedPtr msg){
        counter_+=msg->data; 
    }

    void publishCounter(){
        auto msg = example_interfaces::msg::Int64();
        msg.data = this->counter_;
        publisher_->publish(msg);
    }

    void callBackResetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                            const example_interfaces::srv::SetBool::Response::SharedPtr response){
            if(request->data){
                this->counter_=0;
                response->success = true;
                response->message = "Counter was reset";
            }else{
                response->success = false;
                response->message = "Counter not reset";
            }

    }

    int32_t counter_; 
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;

    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_; 

};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
# #!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool
 
class NumberCounterNode(Node): 
    def __init__(self):
        super().__init__("number_counter") 
        self.counter_ = 0
        self.subscriber_ = self.create_subscription(
            Int64, "number", self.callback_counter, 10
        )
        self.publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.timer_ = self.create_timer(1, self.publish_count)


        self.server_ = self.create_service(
            SetBool, "reset_counter", self.callback_reset_counter)


        self.get_logger().info("Number Counter has started.")


    def callback_counter(self, msg: Int64):
        self.counter_ += msg.data

    def publish_count(self):
        msg = Int64()
        msg.data = self.counter_
        self.publisher_.publish(msg)

    def callback_reset_counter(self, request: SetBool.Request, response: SetBool.Response):
        if request.data: 
            self.counter_ = 0
            response.success = True
            response.message = "Counter successfully reset."
        else:
            response.success = False
            response.message = "Counter not reset."
            
        return response

 
 
def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode() 
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()